healthcheck/healtchcek.go

```diff
@@ -0,0 +1,110 @@
+/*
+ *
+ * Copyright 2017 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package healthcheck
+
+import (
+	"errors"
+	"io"
+	"runtime/debug"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"google.golang.org/grpc/internal/backoff"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
+	"google.golang.org/grpc/internal"
+	"google.golang.org/grpc/status"
+)
+
+func init() {
+	internal.HealthCheckFunc = newClientHealthCheck
+}
+
+const maxDelay time.Duration = 5 * time.Second
+
+func newClientHealthCheck(ctx context.Context, newStream func() (interface{}, error), update func(bool), service string) error {
+	debug.PrintStack()
+	retryCnt := 0
+	doBackoff := false
+	bo := backoff.Exponential{MaxDelay: maxDelay}
+retryConnection:
+	for {
+		if !doBackoff {
```

Comment:
How about putting the retry and exponential backoff logic at the end of the for loop, which I think is clearer? And let the retry logic at the end to decide whether backoff is required. Whenever an error is encountered, just go directly to the retry logic.
```go
backoff  := false
retryCnt := 0
for {
	select{
	case <-ctx.Done():
		return nil
	default:
	}
	
	rawS, err := newStream()
	if err != nil {
		goto retry
	}

	for {
		err = s.RecvMsg(resp)
		if err != nil {
				goto retry
			}
		} else {
			retryCnt = 0
			backoff = false
		}
	}

	retry:
		if !backoff {
			backoff = true
			continue
		}
		timer := time.NewTimer(bo.Backoff(retryCnt++))
		select{
			case <-ctx.Done():
				timer.Stop()
				return nil
			case <timer.C:
		}
}

```